File: TallerOrdenamiento/Heap&Bucket.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <random>

using namespace std;

struct MetricasAlgoritmo {
    long long comparaciones;
    long long intercambios;
    double tiempo_segundos;
};

long long comparaciones_heap = 0;
long long intercambios_heap = 0;

void hundir_nodo_minimo(vector<int>& arreglo, int tamano_heap, int indice_actual) {
    int indice_menor = indice_actual;
    int hijo_izquierdo = 2 * indice_actual + 1;
    int hijo_derecho = 2 * indice_actual + 2;

 
    if (hijo_izquierdo < tamano_heap) {
        comparaciones_heap++;
        if (arreglo[hijo_izquierdo] < arreglo[indice_menor]) {
            indice_menor = hijo_izquierdo;
        }
    }


    if (hijo_derecho < tamano_heap) {
        comparaciones_heap++;
        if (arreglo[hijo_derecho] < arreglo[indice_menor]) {
            indice_menor = hijo_derecho;
        }
    }


    if (indice_menor != indice_actual) {
        swap(arreglo[indice_actual], arreglo[indice_menor]);
        intercambios_heap++;
        hundir_nodo_minimo(arreglo, tamano_heap, indice_menor);
    }
}

MetricasAlgoritmo ejecutar_heapsort_minimos(vector<int> arreglo) {
    comparaciones_heap = 0;
    intercambios_heap = 0;
    int n = arreglo.size();

    auto inicio = chrono::high_resolution_clock::now();

    for (int i = n / 2 - 1; i >= 0; i--) {
        hundir_nodo_minimo(arreglo, n, i);
    }


    for (int i = n - 1; i > 0; i--) {
        swap(arreglo[0], arreglo[i]);
        intercambios_heap++;
        hundir_nodo_minimo(arreglo, i, 0);
    }

    auto fin = chrono::high_resolution_clock::now();
    chrono::duration<double> duracion = fin - inicio;

    return {comparaciones_heap, intercambios_heap, duracion.count()};
}


pair<long long, long long> ordenamiento_insercion_subcubeta(vector<int>& cubeta) {
    long long comp = 0, inter = 0;
    int n = cubeta.size();

    for (int i = 1; i < n; ++i) {
        int valor_actual = cubeta[i];
        int j = i - 1;
        while (j >= 0) {
            comp++;
            if (cubeta[j] > valor_actual) {
                cubeta[j + 1] = cubeta[j];
                inter++;
                j--;
            } else {
                break;
            }
        }
        cubeta[j + 1] = valor_actual;
    }
    return {comp, inter};
}

MetricasAlgoritmo ejecutar_bucket_sort(const vector<int>& datos_originales, int cantidad_cubetas) {
    if (datos_originales.empty()) return {0, 0, 0.0};

    auto inicio = chrono::high_resolution_clock::now();

    int valor_minimo = *min_element(datos_originales.begin(), datos_originales.end());
    int valor_maximo = *max_element(datos_originales.begin(), datos_originales.end());

    if (valor_minimo == valor_maximo) {
        auto fin = chrono::high_resolution_clock::now();
        chrono::duration<double> duracion = fin - inicio;
        return {0, 0, duracion.count()};
    }

    double rango_cubeta = static_cast<double>(valor_maximo - valor_minimo) / cantidad_cubetas;
    vector<vector<int>> arreglo_cubetas(cantidad_cubetas);

    for (int numero : datos_originales) {
        int indice_cubeta = static_cast<int>((numero - valor_minimo) / rango_cubeta);
        if (indice_cubeta >= cantidad_cubetas) {
            indice_cubeta = cantidad_cubetas - 1;
        }
        arreglo_cubetas[indice_cubeta].push_back(numero);
    }

    long long comparaciones_totales = 0;
    long long intercambios_totales = 0;
    vector<int> resultado_final;

    for (auto& cubeta : arreglo_cubetas) {
        auto [comp_sub, inter_sub] = ordenamiento_insercion_subcubeta(cubeta);
        comparaciones_totales += comp_sub;
        intercambios_totales += inter_sub;
        resultado_final.insert(resultado_final.end(), cubeta.begin(), cubeta.end());
    }

    auto fin = chrono::high_resolution_clock::now();
    chrono::duration<double> duracion = fin - inicio;

    return {comparaciones_totales, intercambios_totales, duracion.count()};
}


int main() {

    mt19937 generador(42);
    uniform_int_distribution<int> distribucion(1, 100);
    vector<int> datos_experimento(50);
    for (int& num : datos_experimento) {
        num = distribucion(generador);
    }


    MetricasAlgoritmo metricas_heap = ejecutar_heapsort_minimos(datos_experimento);

    cout << "1. Resultados Heapsort (Min-Heap):" << endl;
    cout << "  • Comparaciones: " << metricas_heap.comparaciones 
         << " | Intercambios: " << metricas_heap.intercambios 
         << " | Tiempo: " << metricas_heap.tiempo_segundos << " s" << endl;

    MetricasAlgoritmo metricas_b5 = ejecutar_bucket_sort(datos_experimento, 5);
    MetricasAlgoritmo metricas_b50 = ejecutar_bucket_sort(datos_experimento, 50);

    cout << "\n2. Resultados Bucket Sort:" << endl;
    cout << "  • Con 5 cubetas  -> Comparaciones: " << metricas_b5.comparaciones 
         << " | Tiempo: " << metricas_b5.tiempo_segundos << " s" << endl;
         
    cout << "  • Con 50 cubetas -> Comparaciones: " << metricas_b50.comparaciones 
         << " | Tiempo: " << metricas_b50.tiempo_segundos << " s" << endl;

    return 0;
}

File: TallerOrdenamiento/CodigosPython/Heap&Bucket.py
import time
import random

comparaciones_heap = 0
intercambios_heap = 0

def hundir_nodo_minimo(lista, tamano_heap, indice_actual):
    global comparaciones_heap, intercambios_heap
    indice_menor = indice_actual
    hijo_izquierdo = 2 * indice_actual + 1
    hijo_derecho = 2 * indice_actual + 2


    if hijo_izquierdo < tamano_heap:
        comparaciones_heap += 1
        if lista[hijo_izquierdo] < lista[indice_menor]:
            indice_menor = hijo_izquierdo

    if hijo_derecho < tamano_heap:
        comparaciones_heap += 1
        if lista[hijo_derecho] < lista[indice_menor]:
            indice_menor = hijo_derecho

    if indice_menor != indice_actual:
        lista[indice_actual], lista[indice_menor] = lista[indice_menor], lista[indice_actual]
        intercambios_heap += 1
        hundir_nodo_minimo(lista, tamano_heap, indice_menor)

def ejecutar_heapsort_minimos(datos_originales):
    global comparaciones_heap, intercambios_heap
    comparaciones_heap = 0
    intercambios_heap = 0
    lista_trabajo = list(datos_originales)
    cantidad_elementos = len(lista_trabajo)

    tiempo_inicio = time.perf_counter()

    for i in range(cantidad_elementos // 2 - 1, -1, -1):
        hundir_nodo_minimo(lista_trabajo, cantidad_elementos, i)

    for i in range(cantidad_elementos - 1, 0, -1):
        lista_trabajo[0], lista_trabajo[i] = lista_trabajo[i], lista_trabajo[0]
        intercambios_heap += 1
        hundir_nodo_minimo(lista_trabajo, i, 0)
        
    tiempo_fin = time.perf_counter()
    duracion_segundos = tiempo_fin - tiempo_inicio
    
    return lista_trabajo, comparaciones_heap, intercambios_heap, duracion_segundos

def ordenamiento_insercion_subcubeta(lista_cubeta):
    comparaciones = 0
    intercambios = 0
    for i in range(1, len(lista_cubeta)):
        valor_actual = lista_cubeta[i]
        j = i - 1
        while j >= 0:
            comparaciones += 1
            if lista_cubeta[j] > valor_actual:
                lista_cubeta[j + 1] = lista_cubeta[j]
                intercambios += 1
                j -= 1
            else:
                break
        lista_cubeta[j + 1] = valor_actual
    return lista_cubeta, comparaciones, intercambios

def ejecutar_bucket_sort(datos_originales, cantidad_cubetas=5):
    if len(datos_originales) == 0:
        return datos_originales, 0, 0, 0
        
    lista_trabajo = list(datos_originales)
    valor_minimo, valor_maximo = min(lista_trabajo), max(lista_trabajo)
    
    if valor_minimo == valor_maximo:
        return lista_trabajo, 0, 0, 0
        
    rango_cubeta = (valor_maximo - valor_minimo) / cantidad_cubetas
    arreglo_cubetas = [[] for _ in range(cantidad_cubetas)]
    
    tiempo_inicio = time.perf_counter()
    
    for numero in lista_trabajo:
        indice_cubeta = int((numero - valor_minimo) / rango_cubeta)
        if indice_cubeta >= cantidad_cubetas:
            indice_cubeta = cantidad_cubetas - 1
        arreglo_cubetas[indice_cubeta].append(numero)
        

    resultado_final = []
    comparaciones_totales = 0
    intercambios_totales = 0
    
    for cubeta in arreglo_cubetas:
        cubeta_ordenada, comp_sub, inter_sub = ordenamiento_insercion_subcubeta(cubeta)
        resultado_final.extend(cubeta_ordenada)
        comparaciones_totales += comp_sub
        intercambios_totales += inter_sub
        
    tiempo_fin = time.perf_counter()
    duracion_segundos = tiempo_fin - tiempo_inicio
    
    return resultado_final, comparaciones_totales, intercambios_totales, duracion_segundos


random.seed(42) 
datos_experimento = [random.randint(1, 100) for _ in range(50)]

heap_resultado, heap_comparaciones, heap_intercambios, heap_tiempo = ejecutar_heapsort_minimos(datos_experimento)

print("Resultados Heapsort (Min-Heap):")
print("  • Lista ordenada:", heap_resultado)
print("  • Comparaciones:", heap_comparaciones)
print("  • Intercambios:", heap_intercambios)
print("  • Tiempo (segundos):", heap_tiempo)

# --- 2. Prueba Bucket Sort (5 vs 50 cubetas) ---
bucket5_resultado, bucket5_comparaciones, bucket5_intercambios, bucket5_tiempo = ejecutar_bucket_sort(datos_experimento, cantidad_cubetas=5)
bucket50_resultado, bucket50_comparaciones, bucket50_intercambios, bucket50_tiempo = ejecutar_bucket_sort(datos_experimento, cantidad_cubetas=50)

print("\nResultados Bucket Sort (5 cubetas):")
print("  • Comparaciones totales:", bucket5_comparaciones)
print("  • Tiempo (segundos):", bucket5_tiempo)

print("\nResultados Bucket Sort (50 cubetas):")
print("  • Comparaciones totales:", bucket50_comparaciones)
print("  • Tiempo (segundos):", bucket50_tiempo)